File: scaffolds/fastapi/app/repositories/base_repository.py
from abc import ABC, abstractmethod
from typing import Any, Generic, TypeVar

T = TypeVar("T")


class BaseRepository(ABC, Generic[T]):
    """
    Abstract base repository. Extend this for each entity.

    Swap the implementation with your actual DB client:
      - SQLAlchemy (async): from sqlalchemy.ext.asyncio import AsyncSession
      - DynamoDB: from aiobotocore.session import AioSession
      - MongoDB: motor.motor_asyncio.AsyncIOMotorClient
    """

    @abstractmethod
    async def find_by_id(self, id: Any) -> T | None:
        raise NotImplementedError

    @abstractmethod
    async def find_all(self, filters: dict[str, Any] | None = None) -> list[T]:
        raise NotImplementedError

    @abstractmethod
    async def create(self, data: dict[str, Any]) -> T:
        raise NotImplementedError

    @abstractmethod
    async def update(self, id: Any, data: dict[str, Any]) -> T | None:
        raise NotImplementedError

    @abstractmethod
    async def delete(self, id: Any) -> bool:
        raise NotImplementedError

File: scaffolds/nodejs/src/dal/base.repository.ts
/**
 * BaseRepository — generic typed contract for data access.
 *
 * Extend this for every entity in your service.
 * Swap the implementation with your actual client:
 *   - DynamoDB:   @aws-sdk/lib-dynamodb  (DynamoDBDocumentClient)
 *   - PostgreSQL: drizzle-orm | prisma | pg
 *   - MongoDB:    mongoose
 */
export abstract class BaseRepository<TEntity, TId = string> {
  protected readonly tableName: string;

  constructor(tableName: string) {
    this.tableName = tableName;
  }

  abstract findById(id: TId): Promise<TEntity | null>;

  abstract findAll(filters?: Partial<TEntity>): Promise<TEntity[]>;

  abstract create(data: Omit<TEntity, 'id' | 'createdAt' | 'updatedAt'>): Promise<TEntity>;

  abstract update(id: TId, data: Partial<TEntity>): Promise<TEntity | null>;

  abstract delete(id: TId): Promise<boolean>;
}

// ─── Example: DynamoDB implementation skeleton ───────────────────────────────
//
// import {
//   DynamoDBClient,
//   GetItemCommand,
//   PutItemCommand,
// } from '@aws-sdk/client-dynamodb';
// import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
//
// export class UserRepository extends BaseRepository<User> {
//   readonly #client: DynamoDBDocumentClient;
//
//   constructor() {
//     super(process.env['TABLE_NAME'] ?? 'users');
//     this.#client = DynamoDBDocumentClient.from(new DynamoDBClient({}));
//   }
//
//   async findById(id: string): Promise<User | null> {
//     const { Item } = await this.#client.send(
//       new GetCommand({ TableName: this.tableName, Key: { pk: id } }),
//     );
//     return (Item as User) ?? null;
//   }
//   ...
// }
